File: src/World/Objects/Clouds.cpp
/*
   Copyright 2010-2011 Tyslenko Max, Bogatirev Pavel.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "World/Objects/Clouds.hpp"
#include "World/Objects/Foreground.hpp"

void Clouds::setLimit(uint16_t limit)
{
   mClouds.resize(limit);

    if (limit > mMaxObjects)
      for (uint16_t i=0; i < limit - mMaxObjects; i++)
          processClouds(mGC, environManager().getWindPower(), mClouds[mMaxObjects + i], false);

   mMaxObjects = limit;
}

void Clouds::init()
{
    float windPower = environManager().getWindPower();
    for (uint16_t i=0; i < mMaxObjects; i++)
    {
        mClouds.push_back(CloudData());
        processClouds(mGC, windPower, mClouds[i], true);
    }
}

void Clouds::processClouds(CL_GraphicContext &gc, float windPower, CloudData &current, bool firstTime)
{
    current.x_speed = 0;
    uint16_t size = resourceManager().sectionHandle("Clouds").get_child_nodes().get_length();
    current.cloudType = rand() % size;
    current.imageHandle = resourceManager().getSprite("Clouds", boost::lexical_cast<std::string>(current.cloudType));

    if (!firstTime)
    {
        if (windPower < 0)
            current.x = mWindowWidth + current.imageHandle.get_width();
        else
            current.x = 0 - current.imageHandle.get_width();

        current.timeout = rand() % 350;
    }
    else
    {
        current.x = rand() % mWindowWidth;
        current.timeout = 0;

        boost::shared_ptr<Foreground> temp = levelManager().getObjectByType<Foreground>("Foreground");
        if (temp->getType() != "Empty") mActualSize = temp->getActualSize();
        current.y_offset = rand() %  mWindowHeight - (mWindowHeight * mActualSize / 100);
    }

    //float alpha = static_cast<float>((rand()%4 + 7)) / 10.0f;
    //float color = static_cast<float>((rand()%1 + 3)) / 10.0f;
    //current.mColor = CL_Colorf(color, color, color, alpha);
    //current.mColor = CL_Colorf(color, color, color, 1.2f);
    current.mColor = CL_Colorf(42/255.0f, 72/255.0f, 85/255.0f, 1.1f);

    current.speed_koef = rand() % 15 + 25;
}

Clouds::Clouds(uint16_t maxClouds)
    : mFirstTime(true)
{
    srand(time(NULL));

    mMaxObjects = maxClouds;
    mGC = appManager().getGraphic();
    mWindowWidth = ScreenResolutionY;
    mWindowHeight = ScreenResolutionX;
    mActualSize = 0;
}

// This function is required by Lightnings (make lightning only from position where
// Cloud is positioned at current time
CL_Pointf Clouds::getCloudPos()
{
    uint16_t counter = 0;
    CL_Pointf result(-1, -1);
    while (result.x == -1)
    {
        ++counter;
        if (counter == 4) break;

        uint16_t index = rand() % (int)mMaxObjects + 1;
        CloudData &current = mClouds[index];
        if (current.x > mWindowWidth*0.1 &&
            current.x < mWindowWidth - mWindowWidth*0.1)
        {
            result.x = current.x;
            result.y = current.y_offset +
                current.imageHandle.get_height() / 2.0f;
        }
        else
        {
            continue;
        }
    }

    return result;
}

void Clouds::setPosition(CL_Pointf newPos)
{
}
CL_Pointf Clouds::getPosition()
{
     return levelManager().getCamViewport().get_top_left();
}
CL_Rectf Clouds::getRectangle()
{
    return levelManager().getCamViewport();
}

void Clouds::update(float elapsed)
{
}

void Clouds::updateVisual(float newX, float newY)
{
    float windPower = environManager().getWindPower();
    float elapsed = appManager().getElapsed() / 1000.0f;
    float newSpeed = windPower * elapsed * 0.5;

    for (uint16_t i=0; i < mMaxObjects; i++)
    {
        CloudData &current = mClouds[i];
        if (current.timeout > 0)
        {
            current.timeout--;
        }
        else
        {
            if (windPower > 0) {
                if (current.x > mWindowWidth)
                    processClouds(mGC, windPower, current);
            } else {
                if (current.x < -current.imageHandle.get_width())
                    processClouds(mGC, windPower, current);
            }

            current.x += current.x_speed * elapsed;
            current.x_speed = current.speed_koef * newSpeed;

            current.imageHandle.set_color(current.mColor);
            current.imageHandle.draw(mGC, current.x, current.y_offset);
        }
    }
}


boost::shared_ptr<Object> Clouds::ParseClouds(CL_DomElement* cloudsElement, std::string& desc)
{
    float maxClouds = 0;
    if(cloudsElement->has_attribute("maxClouds"))
    {
        std::string maxCloudsStr = cloudsElement->get_attribute("maxClouds").c_str();
        maxClouds = boost::lexical_cast<float>(maxCloudsStr);
    }
    return boost::shared_ptr<Object>(new Clouds(maxClouds));
}

File: bin/media/objects/Clouds.py
import Core.LogManager
import Core.ApplicationManager
import Core.WorldManager
import World.Objects.TypesManager
from World.Objects.Object import *
import Core.Utils
import random

class CloudData:
    def __init__(self):
        self.x = random.randint(0, Core.Utils.ScreenResolutionX)
        self.x_speed = self.speed_koef = self.y_offset = 0
        self.cloudType = self.timeout = 0
        self.image = None

class Clouds(Object):
    def __init__(self, maxObjects):
        super(Clouds, self).__init__()

        # We store drop objects here
        self.mClouds = []

        # Precalculated value
        self.mMaxObjects = int(maxObjects)
        self.mCamSpeedOffset = 0

        self.mAppManagerHandle = Core.ApplicationManager.getInstance()
        self.mGC = self.mAppManagerHandle.GetGraphic()
        self.mWorldManagerHandle = Core.WorldManager.getInstance()

    def processClouds(self, windPower, curObject, firsTime):
        size = Core.ResourceManager.SectionNumChildren("Clouds")
        cloud = random.randint(0, size-1)
        curObject.image = Core.ResourceManager.getInstance().GetImage("Clouds", str(cloud))
        color = ColorRGB(42, 72, 85)
        color.SetAlpha(random.uniform(0.9, 1.1))
        curObject.image.SetColor(color)

        scaleK = random.uniform(1.0, 2.0)
        curObject.image.SetScale(scaleK, scaleK)

        if not firsTime:
            if windPower < 0:
                curObject.x = Core.Utils.ScreenResolutionX - curObject.image.GetWidth()
            else:
                curObject.x = 0 - curObject.image.GetWidth()

            curObject.timeout = random.randint(0, 350)
        else:
            curObject.x = random.randint(0, Core.Utils.ScreenResolutionX)
            curObject.timeout = 0

        curObject.y_offset = random.randint(0, Core.Utils.ScreenResolutionY)
        #curObject.speed_koef = random.randint(10, 40)
        curObject.x_speed = random.uniform(0.01, 0.1) * windPower

    def cameraMoved(self, xOffset):
        self.mCamSpeedOffset = xOffset

    def Init(self):
        windPower = self.mWorldManagerHandle.GetWindPower()

        self.mCamera = Core.LevelManager.getInstance().GetCamera()
        self.mCamera.ConnectCameraMoved(self.cameraMoved)

        # Init first rain drops
        for i in xrange(self.mMaxObjects-1):
            self.mClouds.append(CloudData())
            self.processClouds(windPower, self.mClouds[i], True)

    def Update(self, elapsed):
        windPower = self.mWorldManagerHandle.GetWindPower()
        #newSpeed = windPower * elapsed / 1000 * 0.5

        for i in xrange(self.mMaxObjects-1):
            current = self.mClouds[i]
            if current.timeout > 0:
                current.timeout -= 1
            else:
                if current.x > Core.Utils.ScreenResolutionX + current.image.GetWidth() \
                   or current.x + current.image.GetWidth() < -current.image.GetWidth():
                    self.processClouds(windPower, current, False)

                if self.mCamSpeedOffset is not 0:
                    current.x += self.mCamSpeedOffset/2
                else:
                    current.x += current.x_speed * elapsed
        self.mCamSpeedOffset = 0

    def UpdateVisual(self, newX, newY):
        for i in xrange(self.mMaxObjects-1):
            current = self.mClouds[i]
            if current.timeout <= 0:
                current.image.Draw(self.mGC, current.x, current.y_offset)

    def GetPosition(self):
        return CL_Pointf(0, 0)

    def SetPosition(self):
        print 'Setting position'

    def GetRectangle(self):
        return CL_Rectf(0, 0, Core.Utils.ScreenResolutionX, Core.Utils.ScreenResolutionY)

    def CheckCollision(self):
        return False

    @staticmethod
    def ParseClouds(node):
        maxClouds = 0

        if node.HasAttribute("maxClouds"):
            maxClouds = float(node.GetAttribute("maxClouds"))

        area = Core.Utils.Pixels2Meters(Core.Utils.ScreenResolutionX) * \
              Core.Utils.Pixels2Meters(Core.Utils.ScreenResolutionY)

        return Clouds(area * maxClouds)

World.Objects.TypesManager.RegisterParser("CloudsObject", Clouds.ParseClouds)
